components/layout_2020/replaced.rs

```diff
@@ -276,9 +280,72 @@ impl ReplacedContent {
     pub fn make_fragments(
         &self,
         style: &ServoArc<ComputedValues>,
+        containing_block: &ContainingBlock,
         size: PhysicalSize<Au>,
     ) -> Vec<Fragment> {
-        let rect = PhysicalRect::new(PhysicalPoint::origin(), size);
+        let aspect_ratio = self.preferred_aspect_ratio(&containing_block.into(), style);
+        let natural_size = PhysicalSize::new(
+            self.natural_size.width.unwrap_or(size.width),
+            self.natural_size.height.unwrap_or(size.height),
+        );
+
+        let object_fit_size = aspect_ratio
+            .map(|aspect_ratio| {
+                let preserve_aspect_ratio_with_comparison =
+                    |size: PhysicalSize<Au>, comparison: fn(&Au, &Au) -> bool| {
+                        let (width_axis, height_axis) = if !style.writing_mode.is_vertical() {
+                            (Direction::Inline, Direction::Block)
+                        } else {
+                            (Direction::Block, Direction::Inline)
+                        };
+
+                        let candidate_size = PhysicalSize::new(
+                            aspect_ratio.compute_dependent_size(width_axis, size.height),
+                            size.height,
+                        );
+
+                        if comparison(&candidate_size.width, &size.width) {
+                            return candidate_size;
+                        }
+
+                        PhysicalSize::new(
+                            size.width,
+                            aspect_ratio.compute_dependent_size(height_axis, size.width),
+                        )
```

Comment:
We can maybe assert the other comparison

```suggestion
                        let candidate_width =
                            aspect_ratio.compute_dependent_size(width_axis, size.height);
                        if comparison(&candidate_width, &size.width) {
                            return PhysicalSize::new(candidate_width, size.height);
                        }

                        let candidate_height =
                            aspect_ratio.compute_dependent_size(height_axis, size.width);
                        debug_assert!(comparison(&candidate_height, &size.height));
                        PhysicalSize::new(size.width, candidate_height)
```

Reply:
:+1: